packages/hardhat-core/src/internal/hardhat-network/stack-traces/consoleLogger.ts

```diff
@@ -134,7 +134,24 @@ export class ConsoleLogger {
       return;
     }
 
-    return this._decode(parameters, types);
+    const consoleLogs = this._decode(parameters, types);
+
+    this._replaceNumberFormatSpecifiers(consoleLogs);
+
+    return consoleLogs;
+  }
+
+  private _replaceNumberFormatSpecifiers(consoleLogs: ConsoleLogs) {
+    // Replace the occurrences of %d and %i with %s. This is necessary because if the arguments passed are numbers,
+    // they could be too large to be formatted as a Number or an Integer, so it is safer to use a String.
+    // %d and %i are replaced only if there is an odd number of % before the d or i.
+    // If there is an even number of % then it is assumed that the % is escaped and should not be replaced.
+    if (consoleLogs.length > 0) {
+      consoleLogs[0] = String(consoleLogs[0]).replace(
```

Comment:
I always find this `ConsoleLogs` type confusing. Cant's `consoleLogs[0]` be an array?

Or is it safe to cast to string because if the length >0 we are formatting, and hence the first one must be a string?

Reply:
I spent some time on it and I think no, but correct me if I'm wrong because maybe I missed something.

My  findings: the only way that console.log can be used in a .sol file are these:

- 1 argument passed: console.log(123);
- 2 arguments passed: console.log("%d", 123456789123456789123456789);
- more than 2 arguments passed: console.log("%d %%d %%%d %%%%d %%%%%d %%%%%%d", 1, 2, 3);

In all the scenarios the consoleLogs[0]  it's always a single value, not an array, even if multiple console.logs are called at the same time, [see test here](https://github.com/NomicFoundation/hardhat/pull/4115/files/923b72ddceee4cc3e45602a779be78576b883b83#diff-167324083ac14fceebeee729d324066ff06f0eb908f1e6d1cdd25edf2796cd3d).

Reply:
done

Reply:
I think it will always be a string, because the function **_decode**, where I take the consoleLogs, it returns a variable that is of type ConsoleLogs.
ConsoleLogs is this:

<img width="267" alt="image" src="https://github.com/NomicFoundation/hardhat/assets/18092467/50149251-7b31-4493-8f04-1067161559e8">

So string or array of strings ("in a recursive way").

So even if I remove the "String(consoleLogs[0])" in my code to don't force the first element to be a string and I run a test with console.log(1) I still get a "1" instead of a 1.

But yours it's still a good and valid observation, to be 100% safe I could add the check like this:

if (consoleLogs.length > 0 && typeof consoleLogs[0] === "string") {

Reply:
+1 to adding that check, I came here to suggest exactly that before reading this conversation.

Reply:
I agree that the first element is always a string right now, but that could change in the future.

Reply:
Done